### PR TITLE
Refactor `nginx-ingress` Helm chart into component

### PR DESCRIPTION
**Pull request by shafeeqes**
**How to categorize this PR?**

/area dev-productivity quality robustness
/kind technical-debt test
/platform all

**What this PR does / why we need it**:

This PR refactors the deployment for the `nginx-ingress` and adopts the component concept. Consequently, it eliminates the Helm chart and adds full unit tests, instead. This improves several aspects and the overall code quality.

**Which issue(s) this PR fixes**:
Part of #2754

**Special notes for your reviewer**:
CC @rfranzke 

**Release note**:

```other operator
NONE
```

### CODE REVIEW

File: pkg/operation/botanist/component/dependencywatchdog/bootstrap.go [hunk 1]
```diff
@@ -19,16 +19,16 @@ import (
 	"fmt"
 	"time"
 
+	restarterapi "github.com/gardener/dependency-watchdog/pkg/restarter/api"
+	scalerapi "github.com/gardener/dependency-watchdog/pkg/scaler/api"
```
Comment (rfranzke) [inline pkg/operation/botanist/component/dependencywatchdog/bootstrap.go]: Why was this changed/can you move them back to the third group?

File: pkg/operation/botanist/component/nginxingress/nginxingress.go [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	autoscalingv1beta2 "k8s.io/autoscaler/vertical-pod-autoscaler/pkg/apis/autoscaling.k8s.io/v1beta2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	// ManagedResourceName is the name of the ManagedResource containing the resource specifications.
+	ManagedResourceName = "nginx-ingress"
+	// LabelAppValue is the value of a label used for the identification of vpn-shoot pods.
```
Comment (rfranzke) [inline pkg/operation/botanist/component/nginxingress/nginxingress.go]: s/vpn-shoot/nginx-ingress

Btw, why is this constant exposed/is it used anywhere?

File: pkg/operation/botanist/component/nginxingress/nginxingress.go [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	autoscalingv1beta2 "k8s.io/autoscaler/vertical-pod-autoscaler/pkg/apis/autoscaling.k8s.io/v1beta2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	// ManagedResourceName is the name of the ManagedResource containing the resource specifications.
+	ManagedResourceName = "nginx-ingress"
```
Comment (rfranzke) [inline pkg/operation/botanist/component/nginxingress/nginxingress.go]: Why is this constant exposed/is it used anywhere?

File: pkg/operation/botanist/component/nginxingress/nginxingress.go [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+	deploymentName = "nginx-ingress"
+	serviceName    = "nginx-ingress"
+)
+
+// Interface contains functions for a nginxIngress Deployer
+type Interface interface {
+	component.DeployWaiter
+}
+
+// Values is a set of configuration values for the nginxIngress component.
```
Comment (rfranzke) [inline pkg/operation/botanist/component/nginxingress/nginxingress.go]: Please consistently use `nginx-ingress` everywhere in the doc strings.

File: pkg/operation/botanist/component/nginxingress/nginxingress.go [hunk 5]
```diff
@@ -218,11 +344,27 @@ func (n *nginxIngress) computeResourcesData() (map[string][]byte, error) {
...
+func getLabels(resourceType string) map[string]string {
+	labels := map[string]string{
+		v1beta1constants.LabelApp: LabelAppValue,
+		"release":                 "addons",
+	}
+	if resourceType == "controller" {
+		labels[labelKeyComponent] = labelValueController
+	} else {
+		labels[labelKeyComponent] = labelValueBackend
+	}
```
Comment (rfranzke) [inline pkg/operation/botanist/component/nginxingress/nginxingress.go]: Why not simply let the caller pass the value for `labels[labelKeyComponent]` directly?

File: pkg/operation/botanist/component/nginxingress/nginxingress.go [hunk 6]
```diff
@@ -106,26 +119,48 @@ func (n *nginxIngress) Wait(ctx context.Context) error {
...
+
+	utilruntime.Must(kutil.MakeUnique(configMap))
+
 	var (
-		intStrOne = intstr.FromInt(1)
-		registry  = managedresources.NewRegistry(kubernetes.SeedScheme, kubernetes.SeedCodec, kubernetes.SeedSerializer)
+		intStrOne                              = intstr.FromInt(1)
+		uIDPodSecurityContextBackend     int64 = 65534
+		fsGroupPodSecurityContextBackend int64 = 65534
+		uIDPodSecurityContextController  int64 = 101
```
Comment (rfranzke) [inline pkg/operation/botanist/component/nginxingress/nginxingress.go]: Are these constants needed? Can't we simply use `pointer.Int64()`?

File: pkg/operation/botanist/component/nginxingress/nginxingress.go [hunk 7]
```diff
@@ -301,12 +336,60 @@ func (n *nginxIngress) computeResourcesData() (map[string][]byte, error) {
...
+									},
+								},
+								InitialDelaySeconds: 30,
+								TimeoutSeconds:      5,
+							},
+							Ports: []corev1.ContainerPort{{
+								ContainerPort: containerPortBackend,
+								Protocol:      corev1.ProtocolTCP,
+							}},
+							Resources: corev1.ResourceRequirements{
```
Comment (rfranzke) [inline pkg/operation/botanist/component/nginxingress/nginxingress.go:371]: For readability: First requests, second limits.

File: pkg/operation/botanist/component/nginxingress/nginxingress.go [hunk 8]
```diff
@@ -339,12 +422,156 @@ func (n *nginxIngress) computeResourcesData() (map[string][]byte, error) {
...
+													Operator: metav1.LabelSelectorOpIn,
+													Values:   []string{labelValueController},
+												},
+												{
+													Key:      labelKeyRelease,
+													Operator: metav1.LabelSelectorOpIn,
+													Values:   []string{labelValueAddons},
+												},
+											},
+										},
```
Comment (rfranzke) [inline pkg/operation/botanist/component/nginxingress/nginxingress.go]: You can use `MatchLabels` and your `getLabels` function to save some LOC.

File: pkg/operation/botanist/component/nginxingress/nginxingress.go [hunk 9]
```diff
@@ -339,12 +422,156 @@ func (n *nginxIngress) computeResourcesData() (map[string][]byte, error) {
...
+									},
+								},
+							},
+							LivenessProbe: &corev1.Probe{
+								FailureThreshold: 3,
+								Handler: corev1.Handler{
+									HTTPGet: &corev1.HTTPGetAction{
+										Path:   "/healthz",
+										Port:   intstr.FromInt(10254),
```
Comment (rfranzke) [inline pkg/operation/botanist/component/nginxingress/nginxingress.go]: Maybe have a constant for this probe port and reuse it

File: pkg/operation/botanist/component/nginxingress/nginxingress.go [hunk 10]
```diff
@@ -368,3 +595,22 @@ func getLabels(resourceType string) map[string]string {
...
+
+func (n *nginxIngress) getArgs(configMap *corev1.ConfigMap) []string {
+	ingressClass := "seed-nginx"
+	if version.ConstraintK8sGreaterEqual122.Check(n.values.KubernetesVersion) {
+		ingressClass = "k8s.io/seed-nginx"
+	}
+
+	return []string{
+		"/nginx-ingress-controller",
+		"--default-backend-service=garden/nginx-ingress-k8s-backend",
```
Comment (rfranzke) [inline pkg/operation/botanist/component/nginxingress/nginxingress.go]: Use `n.namespace` and the constant for the default backend service name.

File: pkg/operation/botanist/component/nginxingress/nginxingress.go [hunk 11]
```diff
@@ -368,3 +595,22 @@ func getLabels(resourceType string) map[string]string {
...
+		"--enable-ssl-passthrough=true",
+		"--publish-service=garden/nginx-ingress-controller",
```
Comment (rfranzke) [inline pkg/operation/botanist/component/nginxingress/nginxingress.go]: Use `n.namespace` and the constant for the controller service name.

File: pkg/operation/botanist/component/nginxingress/nginxingress.go [hunk 12]
```diff
@@ -368,3 +595,22 @@ func getLabels(resourceType string) map[string]string {
...
+		"--election-id=ingress-controller-seed-leader",
+		`--ingress-class=` + ingressClass,
```
Comment (rfranzke) [inline pkg/operation/botanist/component/nginxingress/nginxingress.go]: Why did you use \` instead of `"`?

File: pkg/operation/botanist/component/nginxingress/nginxingress.go [hunk 13]
```diff
@@ -368,3 +595,22 @@ func getLabels(resourceType string) map[string]string {
...
+		"--update-status=true",
+		"--annotations-prefix=nginx.ingress.kubernetes.io",
+		`--configmap=garden/` + configMap.Name,
```
Comment (rfranzke) [inline pkg/operation/botanist/component/nginxingress/nginxingress.go]: Same here? Also, use `n.namespace`

File: pkg/operation/botanist/component/nginxingress/nginxingress.go [hunk 14]
```diff
@@ -368,3 +595,22 @@ func getLabels(resourceType string) map[string]string {
...
+	}
```
Comment (rfranzke) [inline pkg/operation/botanist/component/nginxingress/nginxingress.go:604]: I don't see the 

```
            {{- if semverCompare ">= 1.22-0" .Capabilities.KubeVersion.GitVersion }}
            - --controller-class={{ include "nginx-ingress.class" . }}
            {{- end }}
```

part which was in the Helm charts. Do I miss something?

File: pkg/operation/botanist/component/nginxingress/nginxingress.go [hunk 15]
```diff
@@ -412,6 +412,9 @@ func (n *nginxIngress) computeResourcesData() (map[string][]byte, error) {
 					v1beta1constants.LabelApp: labelAppValue,
 					labelKeyComponent:         labelValueController,
 				},
+				Annotations: map[string]string{
+					references.AnnotationKey(references.KindConfigMap, configMap.Name): configMap.Name,
```
Comment (rfranzke) [inline pkg/operation/botanist/component/nginxingress/nginxingress.go:394]: Can you add a comment here explaining that the `InjectAnnotations` function is not used since the `ConfigMap` is not mounted as volume (i.e., using the function wouldn't have any effect)?

**Review by rfranzke — CHANGES_REQUESTED** [review]
I think now the `nginx-ingress` component is deployed unconditionally in the seed reconciliation. What about the things in the `seed.go` file (e.g., https://github.com/gardener/gardener/blob/851aebbe5a24c89d1a30801326a976190ddcc54c/pkg/operation/seed/seed.go#L1727-L1737)? The config seems not to be propagated, etc. I haven't checked all things related to nginx-ingress in `seed.go` yet, probably you can do it on your side and adapt the code accordingly.

**Comment by rfranzke** [comment]
The `deleteIngressController` function can now probably be deleted, right?

File: pkg/operation/botanist/component/nginxingress/nginxingress.go (hunk 14 above)
Comment (shafeeqes) [reply pkg/operation/botanist/component/nginxingress/nginxingress.go:604]: I added it later in another commit gardener@973ae11

File: pkg/operation/botanist/component/nginxingress/nginxingress.go (hunk 2 above)
Comment (shafeeqes) [reply pkg/operation/botanist/component/nginxingress/nginxingress.go]: changed in gardener@f650dd9

File: pkg/operation/botanist/component/nginxingress/nginxingress.go (hunk 3 above)
Comment (shafeeqes) [reply pkg/operation/botanist/component/nginxingress/nginxingress.go]: changed in gardener@f650dd9

**Comment by shafeeqes** [comment]
I had to join few commits. It was too messy. I have addressed the review comments in gardener@cda8ad5.
Have some doubts with the seed reconciliation flow. Please take another look. [seed.go](https://github.com/gardener/gardener/pull/5208/files#diff-8495b57b25d39744fe1b20b1e1ce695d130f730142af86ff9c5929afcdbabeb4) . Thanks.
I tested it with seed creation and deletion, It works fine. But I am not really sure, whether this is the right way to do this.

**Comment by shafeeqes** [comment]
We should probably remove this from here, right?
https://github.com/gardener/gardener/blob/799819858891221f1ea7141b7f89ac4e7b4a8786/pkg/operation/seed/seed.go#L1008
if yes, how to refactor it?

we should keep this,
https://github.com/gardener/gardener/blob/799819858891221f1ea7141b7f89ac4e7b4a8786/pkg/operation/seed/seed.go#L1029
right?

**Comment by rfranzke** [comment]
> if yes, how to refactor it?

Yes, it should be removed. You could use the `Values` of the nginx-ingress component to pass the data, similar to how we do it for all other components. Or do you have a more specific question?

> we should keep this,

Yes, for the time being to clean up the old `ConfigMap` since we now migrated to a unique name. Good point. Please add a TODO that we remove this in a future release.

**Comment by shafeeqes** [comment]
Oh, right! Now, I guess I have more stuff to move to the new component, then. Let me work on the seed reconciliation functions once again.

File: pkg/operation/botanist/component/nginxingress/nginxingress.go [hunk 16]
```diff
@@ -0,0 +1,616 @@
+// Copyright (c) 2021 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
```
Comment (rfranzke) [inline pkg/operation/botanist/component/nginxingress/nginxingress.go]: Nit: 2022

File: pkg/operation/botanist/component/nginxingress/nginxingress.go [hunk 17]
```diff
@@ -0,0 +1,616 @@
...
+	// ImageController is the container image used for nginx-ingress controller.
+	ImageController string
+	// ImageDefaultBackend is the container image used for default ingress backend.
+	ImageDefaultBackend string
+	// KubernetesVersion is the version of kubernetes for the seed cluster.
+	KubernetesVersion *semver.Version
+	// IngressClass is the ingress class for the seed nginx-ingress controller
+	IngressClass string
+	// ConfigData contains the configuration details for the nginx-ingress controller
+	ConfigData map[string]interface{}
```
Comment (rfranzke) [inline pkg/operation/botanist/component/nginxingress/nginxingress.go]: This could be `map[string]string` or?

File: pkg/operation/botanist/component/nginxingress/nginxingress.go [hunk 18]
```diff
@@ -0,0 +1,616 @@
...
+}
+
+// InterfaceMapToStringMap converts a map[string]interface{} to map[string]string
+func InterfaceMapToStringMap(in map[string]interface{}) map[string]string {
+	m := make(map[string]string, len(in))
+	for k, v := range in {
+		m[k] = fmt.Sprint(v)
+	}
+	return m
+}
```
Comment (rfranzke) [inline pkg/operation/botanist/component/nginxingress/nginxingress.go]: This function should probably be moved to the `pkg/operation/seed` package. This component can just take `map[string]string` as input

File: pkg/operation/botanist/monitoring.go [hunk 19]
```diff
@@ -404,10 +404,11 @@ func (b *Botanist) DeploySeedGrafana(ctx context.Context) error {
 	return b.deployGrafanaCharts(ctx, common.GrafanaUsersRole, usersDashboards.String(), basicAuthUsers, common.GrafanaUsersPrefix)
 }
 
+//TODO(shafeeqes): Reuse seed.ComputeIngressClass
```
Comment (rfranzke) [inline pkg/operation/botanist/monitoring.go]: What about this TODO?

File: pkg/operation/seed/components.go [hunk 20]
```diff
@@ -87,6 +88,27 @@ func defaultKubeScheduler(c client.Client, imageVector imagevector.ImageVector,
 	return scheduler, nil
 }
 
+func defaultNginxIngress(c client.Client, imageVector imagevector.ImageVector, kubernetesVersion *semver.Version, ingressClass string, config map[string]interface{}) (component.DeployWaiter, error) {
```
Comment (rfranzke) [inline pkg/operation/seed/components.go]: Now that you have a dedicated `nginxingress.go` file, this function could be moved there.

File: pkg/operation/seed/nginxingress.go [hunk 21]
```diff
@@ -0,0 +1,227 @@
+// Copyright (c) 2021 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
```
Comment (rfranzke) [inline pkg/operation/seed/nginxingress.go]: 2022

File: pkg/operation/seed/nginxingress.go [hunk 22]
```diff
@@ -0,0 +1,227 @@
...
+			"worker-processes":             "2",
+		}
+		providerConfig = map[string]interface{}{}
+		err            error
+	)
+	if seed.GetInfo().Spec.Ingress != nil && seed.GetInfo().Spec.Ingress.Controller.ProviderConfig != nil {
+		err = json.Unmarshal(seed.GetInfo().Spec.Ingress.Controller.ProviderConfig.Raw, &providerConfig)
+	}
+
+	return utils.MergeMaps(defaultConfig, providerConfig), err
```
Comment (rfranzke) [inline pkg/operation/seed/nginxingress.go]:
```suggestion
	)

	if seed.GetInfo().Spec.Ingress != nil && seed.GetInfo().Spec.Ingress.Controller.ProviderConfig != nil {
		if err ;= json.Unmarshal(seed.GetInfo().Spec.Ingress.Controller.ProviderConfig.Raw, &providerConfig); err != nil {
			return nil, err
		}
	}

	return utils.MergeMaps(defaultConfig, providerConfig), nil

File: pkg/operation/botanist/monitoring.go (hunk 19 above)
Comment (shafeeqes) [reply pkg/operation/botanist/monitoring.go]: Yes, I'll do it.

**Comment by shafeeqes** [comment]
I tested 
1. Migration of `nginxingress` to managed resource on seeds with v1.20.13 and v1.22.4. found all resources to be consistent. 
2. Creation and deletion of new seed with this branch and didn't face any errors.

File: pkg/operation/botanist/monitoring.go (hunk 19 above)
Comment (shafeeqes) [reply pkg/operation/botanist/monitoring.go]: addressed in gardener@9320053

File: pkg/operation/botanist/component/nginxingress/nginxingress.go (hunk 18 above)
Comment (shafeeqes) [reply pkg/operation/botanist/component/nginxingress/nginxingress.go]: addressed in gardener@65735a2

**Review by rfranzke — DISMISSED** [review]
/lgtm

**Review by rfranzke — APPROVED** [review]
